File: hrms_customize/hrms_customize/report/monthly_attendance_report_bs/monthly_attendance_report_bs.py
# Copyright (c) 2025, nepalsaurav and contributors
# For license information, please see license.txt

# import frappe
import math
from types import SimpleNamespace
from frappe import _
import frappe
from frappe.auth import date_diff
from frappe.email.receive import add_days
from hrms_customize.hrms_customize.doctype.bulk_payroll.bulk_payroll import NEPALI_MONTH_MAP, get_month_end
import nepali_datetime


status_map = {
    "Present": "P",
    "Absent": "A",
    "Half Day/Other Half Absent": "HD/A",
    "Half Day/Other Half Present": "HD/P",
    "Work From Home": "WFH",
    "On Leave": "L",
    "Holiday": "H",
    "Weekly Off": "WO",
}


day_abbr = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]


def execute(filters: dict | None = None):
    """Return columns and data for the report.

    This is the main entry point for the report. It accepts the filters as a
    dictionary and should return columns and data. It is called by the framework
    every time the report is refreshed or a filter is updated.
    """
    columns = get_columns(filters)
    data = get_data(filters)

    return columns, data


def get_columns(filters) -> list[dict]:
    """Return columns for the report.

    One field definition per column, just like a DocType field definition.
    """
    start_date_ad, end_date_ad = get_from_date_and_end_date(filters)
    working_days = date_diff(end_date_ad, start_date_ad) + 1

    columns = [
        {
            "label": _("Employee"),
            "fieldname": "employee",
            "fieldtype": "Data",
        },
        {
            "label": _("Employee Name"),
            "fieldname": "employee_name",
            "fieldtype": "Data",
        },
    ]

    for i in range(working_days):
        day = add_days(start_date_ad, i)
        bs_date = nepali_datetime.date.from_datetime_date(day)
        weekday = day_abbr[day.weekday()]
        label = f"{bs_date.day} {weekday}"

        columns.append({
            "label": label,
            "fieldname": _("{0}").format(bs_date.day),
            "fieldtype": "Data"
        })
    return columns


def get_data(filters) -> list[list]:
    """Return data for the report.

    The report data is a list of rows, with each row being a list of cell values.
    """
    start_date_ad, end_date_ad = get_from_date_and_end_date(filters)

    Attendance = frappe.qb.DocType("Attendance")
    Employee = frappe.qb.DocType("Employee")
    Holiday = frappe.qb.DocType("Holiday")
    attendance_query = (
        frappe.qb.from_(Attendance)
        .join(Employee)
        .on(Employee.name == Attendance.employee)
        .select(
            Attendance.employee,
            Employee.employee_name,
            Attendance.attendance_date,
            Attendance.status,
        )
        .where(
            Attendance.attendance_date.between(
                start_date_ad, end_date_ad)
        )
    )
    attendance_list = attendance_query.run(as_dict=True)

    holiday_query = (
        frappe.qb.from_(Holiday)
        .select(
            Holiday.holiday_date,
            Holiday.description
        )
        .where(
            Holiday.holiday_date.between(start_date_ad, end_date_ad)
        )
    )
    holiday_list = holiday_query.run(as_dict=True)

    holiday_list = [h.holiday_date for h in holiday_list]

    return_dict = {}

    for attendance in attendance_list:
        return_dict[attendance.employee] = {
            "employee": attendance.employee,
            "employee_name": attendance.employee_name
        }

    working_days = date_diff(end_date_ad, start_date_ad) + 1
    for i in range(working_days):
        day = add_days(start_date_ad, i)
        for attendance in attendance_list:
            if attendance.attendance_date == day:
                return_dict[attendance.employee][day] = status_map[attendance.status]
            elif day in holiday_list:
                if day.weekday() == 5:
                    return_dict[attendance.employee][day] = "WO"
                else:
                    return_dict[attendance.employee][day] = "H"
            else:
                if not day in return_dict[attendance.employee]:
                    return_dict[attendance.employee][day] = "A"

    items = []

    for _, value in return_dict.items():
        d = list(value.values())
        items.append(d)

    frappe.log(items)
    return items


def get_from_date_and_end_date(filters):
    month = NEPALI_MONTH_MAP[filters.month]
    year = int(filters.year)
    start_date_bs = nepali_datetime.date(year, month, 1)
    start_date_ad = start_date_bs.to_datetime_date()

    end_date_bs = get_month_end(year, month)
    end_date_ad = end_date_bs.to_datetime_date()
    return start_date_ad, end_date_ad


@frappe.whitelist()
def get_year_month():
    today_date = nepali_datetime.date.today()
    reversed_dict = {v: k for k, v in NEPALI_MONTH_MAP.items()}
    return {
        "year": today_date.year,
        "month": reversed_dict[today_date.month]
    }


@frappe.whitelist()
def convert_nepali_to_ad(date):
    year, month, day = map(int, date.split("-"))
    bs_date = nepali_datetime.date(year, month, day)
    ad_date = bs_date.to_datetime_date().strftime("%Y-%m-%d")
    return ad_date


@frappe.whitelist()
def submit_attendance(year, month):
    start_date_ad, end_date_ad = get_from_date_and_end_date(
        SimpleNamespace(year=year, month=month)
    )

    Attendance = frappe.qb.DocType("Attendance")
    Employee = frappe.qb.DocType("Employee")
    attendance_query = (
        frappe.qb.from_(Attendance)
        .join(Employee)
        .on(Employee.name == Attendance.employee)
        .select(
            Attendance.name,
            Attendance.employee,
            Employee.employee_name,
            Attendance.attendance_date,
            Attendance.status,
        )
        .where(
            Attendance.attendance_date.between(
                start_date_ad, end_date_ad)
        )
    )
    attendance_list = attendance_query.run(as_dict=True)
    length = len(attendance_list)
    for index, value in enumerate(attendance_list):
        attendance = frappe.get_doc("Attendance", value.name)
        try:
            attendance.submit()
        except:
            pass
       
        percent = math.ceil(((index + 1) / length) * 100)
        frappe.publish_progress(percent, title="Submitting...")
    
    frappe.publish_progress(100, title="Submitting...")
    return "success"

File: hrms_customize/hrms_customize/report/monthly_attendance_report_bs/monthly_attendance_report_bs.js
// Copyright (c) 2025, nepalsaurav and contributors
// For license information, please see license.txt

frappe.query_reports["Monthly Attendance Report BS"] = {
	filters: [
		{
			fieldname: "year",
			fieldtype: "Select",
			label: "Year",
			mandatory: 1,
			options:
				"\n2080\n2081\n2082\n2083\n2084\n2085\n2086\n2087\n2088\n2089\n2090\n2091\n2092\n2093\n2094\n2095\n2096\n2097\n2098\n2099",
			wildcard_filter: 0,
		},
		{
			fieldname: "month",
			fieldtype: "Select",
			label: "Month",
			mandatory: 1,
			options:
				"\nBaisakh\nJestha\nAshadh\nShrawan\nBhadra\nAshwin\nKartik\nMangsir\nPoush\nMagh\nFalgun\nChaitra",
			wildcard_filter: 0,
		},
	],

	async onload(report) {
		const r = await frappe.call({
			method: "hrms_customize.hrms_customize.report.monthly_attendance_report_bs.monthly_attendance_report_bs.get_year_month",
		});
	
		report.set_filter_value("year", r.message.year);
		report.set_filter_value("month", r.message.month);


		report.page.add_inner_button("Submit Attendance", async () => {
			frappe.confirm("Are you sure you want to submit attendance, after this you can not edit", () => {
				const year = frappe.query_report.get_filter_value("year")
				const month = frappe.query_report.get_filter_value("month")
				frappe.call({
					method: "hrms_customize.hrms_customize.report.monthly_attendance_report_bs.monthly_attendance_report_bs.submit_attendance",
					args: {year, month},
					callback: (r) => {
						console.log(r)
					}
				})
			})
		}).addClass("btn-primary")
	},

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.colIndex > 3 || column.colIndex > 2) {
			if (value == "HD/P") value = "<span style='color:#914EE3'>" + value + "</span>";
			else if (value == "HD/A") value = "<span style='color:orange'>" + value + "</span>";
			else if (value == "P" || value == "WFH")
				value = "<span style='color:green'>" + value + "</span>";
			else if (value == "A") value = "<span style='color:red'>" + value + "</span>";
			else if (value == "L") value = "<span style='color:#318AD8'>" + value + "</span>";
			else value = "<span style='color:#878787'>" + value + "</span>";
		}
		return value;
	},
};

async function getDefault() {
	const r = await frappe.call({
		method: "hrms_customize.hrms_customize.report.monthly_attendance_report_bs.monthly_attendance_report_bs.get_year_month",
	});
	// console.log(r);
	// return r.message.year
	return "2083";
}

$("<style>")
	.prop("type", "text/css")
	.html(
		`
        .dt-cell {
            cursor: pointer;
        }
    `
	)
	.appendTo("head");

const NEPALI_MONTH_MAP = {
	Baisakh: "01",
	Jestha: "02",
	Ashadh: "03",
	Shrawan: "04",
	Bhadra: "05",
	Ashwin: "06",
	Kartik: "07",
	Mangsir: "08",
	Poush: "09",
	Magh: "10",
	Falgun: "11",
	Chaitra: "12",
};

$(document).on("click", ".dt-cell", async function (event) {
	let parentRow = $(event.target).closest(".dt-row");
	let employee = parentRow.children().eq(1).text().trim();
	let employee_name = parentRow.children().eq(2).text().trim();
	let clickedCell = $(event.target).closest(".dt-cell");
	let cellIndex = clickedCell.index();
	let headerRow = $(".dt-row-header");
	let date = headerRow.children().eq(cellIndex).text().trim();
	if (cellIndex >= 3) {
		console.log(employee, employee_name, date, cellIndex);
		let bs_date = date.split(" ")[0].padStart(2, "0");
		let month = NEPALI_MONTH_MAP[frappe.query_report.get_filter_value("month")];
		bs_date = `${frappe.query_report.get_filter_value("year")}-${month}-${bs_date}`;
		const resp = await frappe.call({
			method: "hrms_customize.hrms_customize.report.monthly_attendance_report_bs.monthly_attendance_report_bs.convert_nepali_to_ad",
			args: { date: bs_date },
		});
		const ad_date = resp.message;

		let dialog = new frappe.ui.Dialog({
			title: "Mark Attendance",
			fields: [
				{
					label: "Employee",
					fieldname: "employee",
					fieldtype: "Link",
					default: employee,
					read_only: 1,
				},
				{
					label: "Employee Name",
					fieldname: "employee_name",
					fieldtype: "Data",
					default: employee_name,
					read_only: 1,
				},
				{
					label: "Attendance Date",
					fieldname: "attendance_date",
					fieldtype: "Date",
					default: ad_date,
					read_only: 1,
				},
				{
					label: "Attendance Date BS",
					fieldname: "attendance_date_bs",
					fieldtype: "Data",
					default: bs_date,
					read_only: 1,
				},
				{
					label: "Status",
					fieldname: "status",
					fieldtype: "Select",
					options: "Present\nAbsent",
					default: "Present",
				},
			],
			size: "small",
			primary_action_label: "Save",
			async primary_action(values) {
				const attendance_record = await frappe.db.get_doc("Attendance", null, {
					"employee": values.employee,
					"attendance_date": values.attendance_date
				})
				await frappe.db.set_value("Attendance", attendance_record.name, "status", values.status)
				dialog.hide();
				frappe.query_report.refresh();
			},
		});

		dialog.show();

		console.log(ad_date);
	}
});
